src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs

```diff
@@ -401,6 +409,115 @@ static void PromoteCommonCommitCharactersOntoList(LSP.VSCompletionList completio
+                    completionHelper,
+                    item,
+                    editorCompletionItem: null,
+                    filterText,
+                    completionTrigger.Kind,
+                    GetFilterReason(completionTrigger),
+                    ImmutableArray<string>.Empty,
+                    includeMatchSpans: false,
+                    ref index,
+                    out var matchResult))
+                {
+                    matchResultsBuilder.Add(matchResult);
+                }
+            }
+
+            // Next, we sort the list based on the pattern matching result.
+            var matchedResults = matchResultsBuilder.ToImmutableArray().Sort(MatchResult<CompletionItem?>.SortingComparer);
+
+            // Finally, truncate the list to 1000 items and any preselected items.
+            var count = MaxCompletionListSize;
+            var filteredList = matchedResults
+                .Where(match => count-- > 0 || match.RoslynCompletionItem.Rules.SelectionBehavior == CompletionItemSelectionBehavior.HardSelection)
+                .Select(match => match.RoslynCompletionItem)
+                .ToImmutableArray();
+            var newCompletionList = completionList.WithItems(filteredList);
+
+            // Per the LSP spec, the completion list should be marked with isIncomplete = false when further insertions will
+            // not generate any more completion items.  This means that we should be checking if the matchedResults is larger
+            // than the filteredList.  However, the VS client has a bug where they do not properly re-trigger completion
+            // when a character is deleted to go from a complete list back to an incomplete list.
+            // For example, the following scenario.
+            // User types "So" -> server gives subset of items for "So" with isIncomplete = true
+            // User types "m" -> server gives entire set of items for "Som" with isIncomplete = false
+            // User deletes "m" -> client has to remember that "So" results were incomplete and re-request if the user types something else, like "n"
+
+            // Currently the VS client does not remember to re-request, so the completion list only ever shows items from "Som"
+            // so we always set the isIncomplete flag to true when the original list size (computed when no filter text was typed) is too large.
+            // VS bug here - https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1335142
+            var isInComplete = completionList.Items.Length > newCompletionList.Items.Length;
+
+            // The list is incomplete, store the result id so we can find this list for followup completion requests.
+            if (isInComplete)
+            {
+                _lastIncompleteResultId = resultId;
+            }
+
+            return (newCompletionList, isInComplete, resultId);
+
+            static CompletionFilterReason GetFilterReason(CompletionTrigger trigger)
```

Comment:
Should this method go into ProtocolConversions?

Reply:
Hmm, I don't think it's necessary for two reasons
1.  it isn't converting between LSP and not LSP (CompletionFilterReason and CompletionTrigger are both roslyn types)
2.  It's very specific to completion list filtering, even if it was converting LSP types imho it can stay here until it is needed in a different handler, it doesn't need to be visible to everything else